torch/csrc/utils/tensor_new.cpp

```diff
@@ -0,0 +1,149 @@
+  auto tensor = type.tensor();
+  tensor.set_(storage);
+  return tensor;
+}
+
+static Tensor new_with_tensor(const Type& type, Tensor other) {
+  if (other.type() != type) {
+    throw TypeError("expected %s (got %s)", type.toString(), other.type().toString());
+  }
+  return other.slice();
+}
+
+static std::vector<int64_t> compute_sizes(PyObject* seq) {
+  std::vector<int64_t> sizes;
+  THPObjectPtr handle;
+  do {
+    auto length = PySequence_Length(seq);
+    if (length < 0) throw python_error();
+    sizes.push_back(length);
+    if (sizes.size() > MAX_DIMS) {
+      throw ValueError("too many dimensions '%s'", Py_TYPE(seq)->tp_name);
+    }
+    if (length == 0) break;
+    handle = THPObjectPtr(PySequence_GetItem(seq, 0));
+    seq = handle.get();
+  } while (PySequence_Check(seq));
+
+  return sizes;
+}
+
+static void recursive_store(char* data, IntList sizes, IntList strides, int64_t dim,
+                            ScalarType scalarType, int elementSize, PyObject* obj) {
+  int64_t ndim = sizes.size();
+  if (dim == ndim) {
+    torch::utils::store_scalar(data, scalarType, obj);
+    return;
+  }
+
+  auto n = sizes[dim];
+  auto seq = THPObjectPtr(PySequence_Fast(obj, "not a sequence"));
+  if (!seq) throw python_error();
+  auto seq_size = PySequence_Fast_GET_SIZE(seq.get());
+  if (seq_size != n) {
+    throw ValueError("expected sequence of length %lld at dim %lld (got %lld)",
+      (long long)n, (long long)dim, (long long)seq_size);
+  }
+
+  PyObject** items = PySequence_Fast_ITEMS(seq.get());
+  for (int64_t i = 0; i < n; i++) {
+    recursive_store(data, sizes, strides, dim + 1, scalarType, elementSize, items[i]);
+    data += strides[dim] * elementSize;
+  }
+}
+
+static Tensor new_from_sequence(ScalarType scalarType, PyObject* data) {
+  if (!PySequence_Check(data)) {
+    throw TypeError("new(): data must be a sequence (got %s)", Py_TYPE(data)->tp_name);
+  }
+  if (THPUtils_checkString(data)) {
+    throw TypeError("new(): invalid data type '%s'", Py_TYPE(data)->tp_name);
+  }
+  if (PyArray_Check(data)) {
+    return autograd::make_variable(tensor_from_numpy(data), false);
+  }
+
+  auto sizes = compute_sizes(data);
+  auto tensor = autograd::make_variable(CPU(scalarType).tensor(sizes), false);
+  recursive_store(
+      (char*)tensor.data_ptr(), tensor.sizes(), tensor.strides(), 0,
+      scalarType, tensor.type().elementSizeInBytes(), data);
+  return tensor;
+}
+
+Tensor tensor_new(const Type& type, PyObject* args, PyObject* kwargs) {
+  static PythonArgParser parser({
+    "new(*, int64_t device=-1)",
+    "new(torch.Size size, *, int64_t device=-1)",
+    "new(Storage storage)",
+    "new(*, int64_t cdata)|hidden",
+    "new(Tensor other)",
+    "new(PyObject* data, *, int64_t device=-1)",
+  });
+  static PythonArgParser parser2({
+    "new(IntList size, *, int64_t device=-1)",
+  });
+
+  PyObject* parsed_args[2];
+
+  // Handle new(int...). This signature would be ambiguous if combined with
```

Comment:
When we have scalars, it won't be great that var.new(5) is equivalent to var.new([5]), not numpy.array(5).  This function already seems overly complex before we even get there -- it seems like we should split it up.  IMO there are really two functions here: size-based and data-based factories (the first two of parser1 and the only of parser2 are size-based, everything else is data-based).  I don't really like these names, but the best I could come up with at the moment are "new" for the size based" and "new_init" for the data-based (i.e. it is equivalent to new + initialization).  I'm sure we can come up with something better.

Reply:
We need to match the semantics of `Tensor.new` to avoid breaking backwards compatibility. We can deprecate some of the overloads, but they all need to be implemented.

Reply:
deprecating some of the overloads sounds like a good solution.  What do you think about having separate size vs data factories?  Which one should "new" be?

Reply:
I think `torch.Tensor()` and `Tensor.new()` should be the size-based constructors.
`torch.tensor()` and (maybe `Tensor.tensor`) should be the data-based factories.

This more-or-less matches numpy where the type `numpy.ndarray` takes a shape and `numpy.array` takes data.

I'd like to get rid of the the other constructors:
`torch.Tensor(tensor)` is equivalent to `tensor[:]`
`torch.Tensor(_cdata=...)` could be `torch._unsafe_from_cdata` or something. (I'm not even sure we use it anymore.
`torch.Tensor(storage)` could be `torch.from_storage(storage)` or we could get rid of storages from the Python API.